###**Project - Omdena Retrolux Challenge**

#### Description of Code:
1) To populate the annual hours for projects with zero and null annual hours using the daily hours and weeks per year.

2) To add address, city, state, zipcodes, longitude and latitude to Features dataset.

### Import necessary packages and datasets

In [1]:
import pandas as pd 
# Mount Google Drive to Collaboratory 
from google.colab import drive
drive.mount('/content/gdrive')


Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [2]:
# Read dataset into dataframes
df_features = pd.read_csv('/content/gdrive/MyDrive/Colab Notebooks/Final_complete.csv')
df_address = pd.read_csv('/content/gdrive/MyDrive/Colab Notebooks/Addresses_Geodata_Projects Merge - Sheet1.csv')

In [3]:
df_features.shape

(5425, 49)

#### Replace populate zero and null annual hours with daily hours and weeks per year

In [4]:
# Check number of rows with null annual hours 
df_features[df_features['ave_annual_hours'].isnull()].shape


(641, 49)

There are 641 rows with null annual hours 

In [5]:
# Check number of rows with zero annual hours 
df_features[df_features['ave_annual_hours']==0].shape

(1, 49)

Just one row has zero annual hours.


In [6]:
# Create a new column called average weekly hours as the sum of all the average daily hours
df_features ['ave_weekly_hours']=df_features ['monday_ave_daily_hours']+df_features ['tuesday_ave_daily_hours']+df_features ['wednesday_ave_daily_hours']+df_features ['thursday_ave_daily_hours']+df_features ['friday_ave_daily_hours']+df_features ['saturday_ave_daily_hours']+df_features ['sunday_ave_daily_hours']
df_features.head()

Unnamed: 0  project_id  option_id  ...  payback         ROI ave_weekly_hours
0           0           3        8.0  ...      NaN    0.000000              0.0
1           1           8      290.0  ...     0.34  297.198880              0.0
2           2          11       92.0  ...      NaN    0.000000              NaN
3           3          15       21.0  ...      NaN         inf              0.0
4           4          15      928.0  ...      NaN   12.304965              0.0

[5 rows x 50 columns]

In [7]:
# Replace the null annual hours with the product of weekly hours and weeks_per_year
df_features.loc[(df_features['ave_weekly_hours']>=0)& (df_features['ave_annual_hours'].isnull()),'ave_annual_hours']=df_features['ave_weekly_hours']*df_features['ave_weeks_per_year']
df_features[df_features['ave_annual_hours'].isnull()].shape

(268, 50)

The number of rows with null annual hours has reduced from 641 to 268.

In [8]:
df_features[df_features['ave_annual_hours'].isnull()]['ROI'].value_counts()

0.000000      245
0.361474        1
429.653740      1
Name: ROI, dtype: int64

The 268 rows also have null values in their weekly hours and weeks per year. Most of them also have zero ROI

In [9]:
# Replace the zero annual hours with the product of weekly hours and weeks_per_year
df_features.loc[(df_features['ave_weekly_hours']>=0)& (df_features['ave_annual_hours']==0),'ave_annual_hours']=df_features['ave_weekly_hours']*df_features['ave_weeks_per_year']
df_features[df_features['ave_annual_hours']==0].shape

(0, 50)

The only row with zero annual hours has been populated with the product of weekly hours and weeks per year.

In [10]:
df_features.head()

Unnamed: 0  project_id  option_id  ...  payback         ROI ave_weekly_hours
0           0           3        8.0  ...      NaN    0.000000              0.0
1           1           8      290.0  ...     0.34  297.198880              0.0
2           2          11       92.0  ...      NaN    0.000000              NaN
3           3          15       21.0  ...      NaN         inf              0.0
4           4          15      928.0  ...      NaN   12.304965              0.0

[5 rows x 50 columns]

In [11]:
df_features.columns

Index(['Unnamed: 0', 'project_id', 'option_id', 'option_calculation_id',
       'company_id', 'projects_name', 'facility_type_id', 'facility_type',
       'industry_type_id', 'name', 'started_at', 'finished_at',
       'tax_rate_source', 'finance_discount_rate', 'tax_exempt', 'active',
       'area_count', 'area_count_sum', 'ave_annual_hours',
       'monday_ave_daily_hours', 'tuesday_ave_daily_hours',
       'wednesday_ave_daily_hours', 'thursday_ave_daily_hours',
       'friday_ave_daily_hours', 'saturday_ave_daily_hours',
       'sunday_ave_daily_hours', 'ave_weeks_per_year', 'hour_type',
       'max_kwh_cost', 'max_kwh_cost_simple', 'existing_lighting_count',
       'total_watts_per_product', 'ave_lm70', 'total_solution_count',
       'total_existing_quantity', 'total_proposed_quantity',
       'total_replacement_count', 'total_fixture_watts',
       'total_fixture_watt_hours', 'total_fixture_life_years',
       'total_item_watts', 'sol_total_existing_count', 'total_product_count',

In [12]:
# drop the daily average hours, the weekly hours and weeks_per_year columns.
df_features.drop(columns = ['monday_ave_daily_hours', 'tuesday_ave_daily_hours','wednesday_ave_daily_hours', 'thursday_ave_daily_hours','friday_ave_daily_hours', 'saturday_ave_daily_hours',
       'sunday_ave_daily_hours', 'ave_weeks_per_year', 'hour_type','ave_weekly_hours'], inplace = True)
df_features.shape

(5425, 40)

#### Add Columns for Address, City, State, and Zipcode

In [13]:
df_address.head()

created_by_user_id  active  ...  Projects.custom_attributes Projects.measure_types
0                  40    True  ...                          {}             {lighting}
1                  40    True  ...                          {}             {lighting}
2                  40    True  ...                          {}             {lighting}
3                  40    True  ...                          {}             {lighting}
4                1300    True  ...                          {}             {lighting}

[5 rows x 75 columns]

In [14]:
# create a smaller dataframe with the relevant columns we need
df = df_address[['Projects.id','address','city','state','zip_code','geodata.latitude','geodata.longitude']]
df.rename(columns = {'Projects.id':'project_id','geodata.latitude':'latitude','geodata.longitude':'longitude'}, inplace = True ) # rename the project_id foreign to facilitate merge
df.head()

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


project_id             address           city  ... zip_code  latitude  longitude
0        8433  111 S Franklin Ave  Valley Stream  ...    11580  40.68510  -73.70180
1        6784   1150 W Lacey Blvd        Hanford  ...    93230  36.34356 -119.70874
2        6784   1150 W Lacey Blvd        Hanford  ...    93230  36.40134 -119.71874
3        6784   1150 W Lacey Blvd        Hanford  ...    93230  36.30328 -119.63624
4        6277       1900 Hazel St     Pine Bluff  ...    71603  34.18065  -92.12348

[5 rows x 7 columns]

In [15]:
# Merge the data set with our features data
df_features_final = pd.merge(df_features,df,on = 'project_id', how = 'left')
df_features_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5625 entries, 0 to 5624
Data columns (total 46 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                5625 non-null   int64  
 1   project_id                5625 non-null   int64  
 2   option_id                 5625 non-null   float64
 3   option_calculation_id     5625 non-null   float64
 4   company_id                5625 non-null   int64  
 5   projects_name             5625 non-null   object 
 6   facility_type_id          2585 non-null   float64
 7   facility_type             2585 non-null   object 
 8   industry_type_id          2538 non-null   float64
 9   name                      5625 non-null   object 
 10  started_at                5625 non-null   object 
 11  finished_at               5625 non-null   object 
 12  tax_rate_source           5610 non-null   float64
 13  finance_discount_rate     5625 non-null   int64  
 14  tax_exem

Only 483 rows were populated with address, city, state and zip_code information and 356 rows with longitude and latitude information

In [16]:
df_features_final.shape

(5625, 46)

In [17]:
df_features_final.to_csv (r'/content/gdrive/MyDrive/Colab Notebooks/features_complete_final.csv', index = False, header=True)